File: 1382-balance-a-binary-search-tree/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
    var node_list []int
    traverse(root, &node_list)
    return buildBST(0, len(node_list)-1, node_list)
}

func traverse(node *TreeNode, node_list *[]int) {
    if node == nil {
        return
    }
    traverse(node.Left, node_list)
    *node_list = append(*node_list, node.Val)
    traverse(node.Right, node_list)
}

func buildBST(left, right int, node_list []int) *TreeNode {
    if left > right {
        return nil
    }
    mid := left + (right-left)/2
    root := &TreeNode{Val: node_list[mid]}
    root.Left = buildBST(left, mid-1, node_list)
    root.Right = buildBST(mid+1, right, node_list)
    return root
}

File: 1382-balance-a-binary-search-tree/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        
        node_list: list[int] = []

        def traverse(node: TreeNode):
            if not node: return

            traverse(node.left)
            node_list.append(node.val)
            traverse(node.right)

        def build_bst(left: int, right: int) -> TreeNode:
            if left > right: return None
            mid: int = left + (right - left) // 2
            root = TreeNode(node_list[mid])
            root.left = build_bst(left, mid-1)
            root.right = build_bst(mid + 1, right)
            return root
        
        traverse(root)
        return build_bst(0, len(node_list) - 1)
            
